rpc/RPCTransaction.java

```diff
@@ -165,12 +157,19 @@ public void close() {
             final UUID requestId = UUID.randomUUID();
             request.setId(requestId.toString());
             request.setLatencyMillis(networkLatencyMillis);
-            collectors.put(requestId, responseCollector);
+            assertOpenAndAddCollector(requestId, responseCollector);
             final QueryIterator<T> queryIterator = new QueryIterator<>(request.build(), requestObserver, responseCollector, transformResponse);
             return StreamSupport.stream(((Iterable<T>) () -> queryIterator).spliterator(), false);
         }
     }
 
+    private void assertOpenAndAddCollector(UUID requestId, ResponseCollector collector) {
+        collectors.compute(requestId, (id, col) -> {
+            if (transactionWasClosed.get()) throw new GraknClientException(TRANSACTION_CLOSED);
+            return collector;
+        });
```

Comment:
Here we atomically verify that the transaction has not been closed before retrieving a `ResponseCollector` from our `ConcurrentMap`.